e2e/databases/spanner.go

```diff
@@ -0,0 +1,184 @@
+package databases
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	instances "cloud.google.com/go/spanner/admin/instance/apiv1"
+	"github.com/go-logr/logr"
+	"github.com/nightlyone/lockfile"
+
+	//revive:disable:dot-imports convention is dot-import
+	. "github.com/onsi/gomega"
+	"google.golang.org/api/option"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+	"google.golang.org/genproto/googleapis/spanner/admin/instance/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	e2eutil "github.com/authzed/spicedb-operator/e2e/util"
+)
+
+type SpannerProvider struct {
+	kclient    kubernetes.Interface
+	namespace  string
+	lock       lockfile.Lockfile
+	mapper     meta.RESTMapper
+	restConfig *rest.Config
+}
+
+func NewSpannerProvider(mapper meta.RESTMapper, restConfig *rest.Config, namespace string) *SpannerProvider {
+	k, err := kubernetes.NewForConfig(restConfig)
+	Expect(err).To(Succeed())
+	lockPath, err := filepath.Abs("./spanner.lck")
+	Expect(err).To(Succeed())
+	lock, err := lockfile.New(lockPath)
+	Expect(err).To(Succeed())
+	return &SpannerProvider{kclient: k, namespace: namespace, lock: lock, mapper: mapper, restConfig: restConfig}
+}
+
+func (p *SpannerProvider) New(ctx context.Context) *LogicalDatabase {
+	p.ensureDatabase(ctx)
+	var logicalDBName string
+
+	Eventually(func(g Gomega) {
+		logicalDBName = names.SimpleNameGenerator.GenerateName("sp")
+		p.execDatabase(ctx, g, func(client *database.DatabaseAdminClient) {
+			op, err := client.CreateDatabase(ctx, &adminpb.CreateDatabaseRequest{
+				Parent:          "projects/fake-project-id/instances/fake-instance",
+				CreateStatement: "CREATE DATABASE `" + logicalDBName + "`",
+			})
+			g.Expect(err).To(Succeed())
+
+			_, err = op.Wait(ctx)
+			g.Expect(err).To(Succeed())
+		})
+	}).Should(Succeed())
+	return &LogicalDatabase{
+		DatastoreURI: fmt.Sprintf("projects/fake-project-id/instances/fake-instance/databases/%s", logicalDBName),
+		DatabaseName: logicalDBName,
+		ExtraConfig: map[string]string{
+			"datastoreSpannerEmulatorHost": fmt.Sprintf("spanner-service.%s:9010", p.namespace),
+		},
+		Engine: "spanner",
+	}
+}
+
+func (p *SpannerProvider) Cleanup(ctx context.Context, db *LogicalDatabase) {
+	// TODO: figure out how to cleanup a spanner emulator db
+}
+
+func (p *SpannerProvider) execInstance(ctx context.Context, g Gomega, cmd func(client *instances.InstanceAdminClient)) {
+	ctx, cancel := context.WithTimeout(ctx, 500*time.Second)
```

Comment:
500 seconds would seem like a lot to execute a statement in the instance, whereas we are having a shorter timeout for other DSs

Reply:
hmm, not sure where that came from. it does seem really high, and I don't remember having a particular issue with spanner. I'll lower it